addOns/client/src/main/java/org/zaproxy/addon/client/internal/ClientMapWriter.java

```diff
@@ -0,0 +1,181 @@
+        // simple - it still uses jackson for value output
+        String indent = " ".repeat(level * 2);
+
+        outputKV(
+                fw,
+                indent,
+                true,
+                NODE_KEY,
+                level == 0 ? ROOT_NODE_NAME : node.getUserObject().getName());
+
+        if (node.getUserObject().isStorage()) {
+            outputKV(fw, indent, false, STORAGE_KEY, node.getUserObject().isStorage());
+        }
+        if (!node.getUserObject().isVisited()) {
+            outputKV(fw, indent, false, VISITED_KEY, node.getUserObject().isVisited());
+        }
+        if (node.getUserObject().getComponents() != null
+                && !node.getUserObject().getComponents().isEmpty()) {
+            for (ClientSideComponent component : node.getUserObject().getComponents()) {
+                if (component.getType() == ClientSideComponent.Type.REDIRECT) {
+                    outputKV(
+                            fw, indent, false, component.getType().getLabel(), component.getHref());
+                } else if (component.getType() == ClientSideComponent.Type.CONTENT_LOADED) {
+                    outputKV(fw, indent, false, component.getType().getLabel(), true);
+                }
+            }
+            indent = outputComponents(fw, node.getUserObject().getComponents(), level, indent);
+        }
+
+        Stats.incCounter(ExtensionClientIntegration.PREFIX + ".export.clientmap.node");
+
+        if (node.getChildCount() > 0) {
+            fw.write(indent);
+            fw.write("  ");
+            fw.write(CHILDREN_KEY);
+            fw.write(":");
+            fw.write('\n');
+            node.children()
+                    .asIterator()
+                    .forEachRemaining(
+                            c -> {
+                                try {
+                                    outputNode(fw, (ClientNode) c, level + 1);
+                                } catch (IOException e) {
+                                    throw new UncheckedIOException(e);
+                                }
+                            });
+        }
+    }
+
+    private static String outputComponents(
+            Writer fw, Set<ClientSideComponent> components, int level, String indent)
+            throws IOException {
+        fw.write(indent);
+        fw.write("  ");
+        fw.write(COMPONENTS_KEY);
+        fw.write(":\n");
+
+        indent = " ".repeat((level + 1) * 2);
+        SortedSet<ClientSideComponent> sortedComponents = new TreeSet<>(components);
+
+        for (ClientSideComponent component : sortedComponents) {
+            boolean first = true;
+            if ((component.getType() != null
+                    && COMPONENT_TYPES_TO_SKIP.contains(component.getType()))) {
+                continue;
+            }
+            first = outputKV(fw, indent, first, NODE_TYPE_KEY, component.getType().getLabel());
```

Comment:
SHould I add a block here to prevent the NPE @psiinon was seeing even though I couldn't recreate it?

```java
if (component.getType == null) {
    continue;
}
```

Maybe with some sort of DEBUG message, and actually maybe first thing in the loop. (Then remove the not null check along with the types to skip condition).

Reply:
With the constructor requiring a Type this should no longer be an issue.